javalib/src/main/scala/java/lang/Integer.scala

```diff
@@ -82,46 +115,88 @@ object Integer {
+
   @inline
   private def parseIntImpl(s: String, radix: scala.Int,
       signed: scala.Boolean): scala.Int = {
 
-    def fail(): Nothing =
-      throw new NumberFormatException(s"""For input string: "$s"""")
+    def fail(): Nothing = parseIntFail(s)
 
-    val len = if (s == null) 0 else s.length
-
-    if (len == 0 || radix < Character.MIN_RADIX || radix > Character.MAX_RADIX)
+    if (s == null || s == "")
       fail()
 
     val firstChar = s.charAt(0)
     val negative = signed && firstChar == '-'
-
-    val maxAbsValue: scala.Double = {
-      if (!signed) 0xffffffffL.toDouble
-      else if (negative) 0x80000000L.toDouble
-      else 0x7fffffffL.toDouble
+    val start = if (negative || firstChar == '+') 1 else 0
+
+    val unsignedResult = parseIntInternal(s, radix, start)
+
+    if (signed) {
+      if (negative) {
+        val result = -unsignedResult
+        if (result > 0)
+          fail()
+        result
+      } else {
+        if (unsignedResult < 0)
+          fail()
+        unsignedResult
+      }
+    } else {
+      unsignedResult
     }
+  }
+
+  @noinline
+  private def parseIntInternal(s: String, radix: Int, start: Int): Int = {
+    def fail(): Nothing = parseIntFail(s)
 
-    var i = if (negative || firstChar == '+') 1 else 0
+    val len = s.length()
 
-    // We need at least one digit
-    if (i >= s.length)
+    // We need at least one digit, even if it is a 0 digit; also check the radix
+    if (start == len || radix < Character.MIN_RADIX || radix > Character.MAX_RADIX)
       fail()
 
-    var result: scala.Double = 0.0
-    while (i != len) {
+    val radixInfo = StringRadixInfos(radix)
+    val maxLength = radixInfo.maxLength
+
+    // Skip over leading 0 digits
+    var i = start
+    while (i != len && Character.isZeroDigit(s.charAt(i)))
+      i += 1
+
+    // Check that the remaining string is not too long
+    val remainingLength = len - i
+    if (remainingLength > maxLength)
+      fail()
+
+    // Until which index can we safely loop without worrying about overflow?
+    val safeLen = if (remainingLength == maxLength) len - 1 else len
+
+    var result: Int = 0
+
+    // Loop until safeLen, during which no overflow can occur
+    while (i != safeLen) {
       val digit = Character.digitWithValidRadix(s.charAt(i), radix)
-      result = result * radix + digit
-      if (digit == -1 || result > maxAbsValue)
+      if (digit == -1)
         fail()
+      result = result * radix + digit
       i += 1
     }
 
-    if (negative)
-      asInt(-result)
-    else
-      asInt(result)
+    // Possibly handle the last digit, where we have to check for overflow
+    if (safeLen != len) {
+      val digit = Character.digitWithValidRadix(s.charAt(safeLen), radix)
+      if (digit == -1 || (result ^ SignBit) > (radixInfo.overflowBarrier ^ SignBit))
+        fail()
+      result = result * radix + digit
+      if ((result ^ SignBit) < (Character.MAX_RADIX ^ SignBit)) // unsigned comparison
+        fail()
```

Comment:
Could you amend the test suite to have a failure case for this case and the overflow barrier case above? (IIUC this is not the case right now).

Reply:
I added comments to existing test cases that trigger these particular overflow checks.